File: packages/shared-types/src/reevaluation.ts
/**
 * Re-evaluation Assessment types for PhysioFlow EMR
 *
 * Re-evaluations compare current assessment values to baseline (initial assessment)
 * values, tracking improvements and declines for ROM, MMT, and outcome measures.
 */

import { z } from 'zod';

// -----------------------------------------------------------------------------
// Enums
// -----------------------------------------------------------------------------

export type ReevaluationMeasureType = 'rom' | 'mmt' | 'outcome_measure';

export type InterpretationResult = 'improved' | 'declined' | 'stable';

// -----------------------------------------------------------------------------
// Core Types
// -----------------------------------------------------------------------------

export interface ReevaluationAssessment {
  id: string;
  patientId: string;
  visitId?: string;
  clinicId: string;
  baselineAssessmentId?: string;
  assessmentType: ReevaluationMeasureType;
  measureLabel: string;
  currentValue: number;
  baselineValue: number;
  change: number;
  changePercentage?: number;
  higherIsBetter: boolean;
  mcidThreshold?: number;
  mcidAchieved: boolean;
  interpretation: InterpretationResult;
  therapistId: string;
  notes?: string;
  assessedAt: string;
  createdAt: string;
  updatedAt: string;
}

export interface ComparisonResult {
  measureLabel: string;
  assessmentType: ReevaluationMeasureType;
  baselineValue: number;
  currentValue: number;
  change: number;
  changePercentage?: number;
  higherIsBetter: boolean;
  mcidThreshold?: number;
  mcidAchieved: boolean;
  interpretation: InterpretationResult;
}

export interface ReevaluationSummary {
  patientId: string;
  visitId?: string;
  therapistId: string;
  assessedAt: string;
  comparisons: ReevaluationAssessment[];
  totalItems: number;
  improved: number;
  declined: number;
  stable: number;
  mcidAchieved: number;
}

// -----------------------------------------------------------------------------
// Request Types
// -----------------------------------------------------------------------------

export interface CreateReevaluationItemRequest {
  assessmentType: ReevaluationMeasureType;
  measureLabel: string;
  currentValue: number;
  baselineValue: number;
  higherIsBetter: boolean;
  mcidThreshold?: number;
}

export interface CreateReevaluationRequest {
  patientId: string;
  visitId?: string;
  baselineAssessmentId?: string;
  assessments: CreateReevaluationItemRequest[];
  notes?: string;
  assessedAt?: string;
}

// -----------------------------------------------------------------------------
// Zod Schemas
// -----------------------------------------------------------------------------

export const ReevaluationMeasureTypeSchema = z.enum(['rom', 'mmt', 'outcome_measure']);

export const InterpretationResultSchema = z.enum(['improved', 'declined', 'stable']);

export const CreateReevaluationItemSchema = z.object({
  assessmentType: ReevaluationMeasureTypeSchema,
  measureLabel: z.string().min(1).max(120),
  currentValue: z.number(),
  baselineValue: z.number(),
  higherIsBetter: z.boolean(),
  mcidThreshold: z.number().positive().optional(),
});

export const CreateReevaluationSchema = z.object({
  patientId: z.string().uuid(),
  visitId: z.string().uuid().optional(),
  baselineAssessmentId: z.string().uuid().optional(),
  assessments: z.array(CreateReevaluationItemSchema).min(1),
  notes: z.string().max(2000).optional(),
  assessedAt: z.string().datetime().optional(),
});

export const ReevaluationAssessmentSchema = z.object({
  id: z.string().uuid(),
  patientId: z.string().uuid(),
  visitId: z.string().uuid().optional(),
  clinicId: z.string().uuid(),
  baselineAssessmentId: z.string().uuid().optional(),
  assessmentType: ReevaluationMeasureTypeSchema,
  measureLabel: z.string(),
  currentValue: z.number(),
  baselineValue: z.number(),
  change: z.number(),
  changePercentage: z.number().optional(),
  higherIsBetter: z.boolean(),
  mcidThreshold: z.number().optional(),
  mcidAchieved: z.boolean(),
  interpretation: InterpretationResultSchema,
  therapistId: z.string().uuid(),
  notes: z.string().optional(),
  assessedAt: z.string().datetime(),
  createdAt: z.string().datetime(),
  updatedAt: z.string().datetime(),
});

// -----------------------------------------------------------------------------
// Utility: Interpretation logic (mirrors Go implementation)
// -----------------------------------------------------------------------------

/**
 * Calculate the change and percentage between baseline and current values.
 */
export function calculateChange(
  baseline: number,
  current: number
): { change: number; changePercentage?: number } {
  const change = current - baseline;
  let changePercentage: number | undefined;
  if (Math.abs(baseline) > 0.0001) {
    changePercentage = (change / Math.abs(baseline)) * 100;
  }
  return { change, changePercentage };
}

/**
 * Determine clinical interpretation of change relative to MCID.
 */
export function determineInterpretation(
  change: number,
  mcidThreshold: number | undefined,
  higherIsBetter: boolean
): { interpretation: InterpretationResult; mcidAchieved: boolean } {
  const threshold = mcidThreshold && mcidThreshold > 0 ? mcidThreshold : 0;
  const absChange = Math.abs(change);

  // If there is a meaningful MCID and the change does not exceed it, consider stable
  if (threshold > 0 && absChange < threshold) {
    return { interpretation: 'stable', mcidAchieved: false };
  }

  // If there is no MCID but the change is essentially zero, consider stable
  if (threshold === 0 && absChange < 0.0001) {
    return { interpretation: 'stable', mcidAchieved: false };
  }

  const mcidAchieved = threshold > 0 && absChange >= threshold;

  if (higherIsBetter) {
    return {
      interpretation: change > 0 ? 'improved' : 'declined',
      mcidAchieved,
    };
  }

  // For pain scales where lower is better
  return {
    interpretation: change < 0 ? 'improved' : 'declined',
    mcidAchieved,
  };
}

File: apps/api/internal/model/reevaluation.go
package model

import (
	"math"
	"time"
)

// AssessmentType represents the type of clinical assessment being re-evaluated.
type AssessmentType string

const (
	AssessmentTypeROM            AssessmentType = "rom"
	AssessmentTypeMMT            AssessmentType = "mmt"
	AssessmentTypeOutcomeMeasure AssessmentType = "outcome_measure"
)

// Interpretation represents the clinical interpretation of change.
type Interpretation string

const (
	InterpretationImproved Interpretation = "improved"
	InterpretationDeclined Interpretation = "declined"
	InterpretationStable   Interpretation = "stable"
)

// ReevaluationAssessment represents a single re-evaluation comparison record.
type ReevaluationAssessment struct {
	ID                   string         `json:"id" db:"id"`
	PatientID            string         `json:"patient_id" db:"patient_id" validate:"required,uuid"`
	VisitID              *string        `json:"visit_id,omitempty" db:"visit_id"`
	ClinicID             string         `json:"clinic_id" db:"clinic_id" validate:"required,uuid"`
	BaselineAssessmentID *string        `json:"baseline_assessment_id,omitempty" db:"baseline_assessment_id"`
	AssessmentType       AssessmentType `json:"assessment_type" db:"assessment_type" validate:"required"`
	MeasureLabel         string         `json:"measure_label" db:"measure_label" validate:"required,max=120"`
	CurrentValue         float64        `json:"current_value" db:"current_value"`
	BaselineValue        float64        `json:"baseline_value" db:"baseline_value"`
	Change               float64        `json:"change" db:"change"`
	ChangePercentage     *float64       `json:"change_percentage,omitempty" db:"change_percentage"`
	HigherIsBetter       bool           `json:"higher_is_better" db:"higher_is_better"`
	MCIDThreshold        *float64       `json:"mcid_threshold,omitempty" db:"mcid_threshold"`
	MCIDAchieved         bool           `json:"mcid_achieved" db:"mcid_achieved"`
	Interpretation       Interpretation `json:"interpretation" db:"interpretation"`
	TherapistID          string         `json:"therapist_id" db:"therapist_id" validate:"required,uuid"`
	Notes                string         `json:"notes,omitempty" db:"notes"`
	AssessedAt           time.Time      `json:"assessed_at" db:"assessed_at"`
	CreatedAt            time.Time      `json:"created_at" db:"created_at"`
	UpdatedAt            time.Time      `json:"updated_at" db:"updated_at"`
}

// ComparisonResult encapsulates the computed comparison between baseline and current values.
type ComparisonResult struct {
	MeasureLabel     string         `json:"measure_label"`
	AssessmentType   AssessmentType `json:"assessment_type"`
	BaselineValue    float64        `json:"baseline_value"`
	CurrentValue     float64        `json:"current_value"`
	Change           float64        `json:"change"`
	ChangePercentage *float64       `json:"change_percentage,omitempty"`
	HigherIsBetter   bool           `json:"higher_is_better"`
	MCIDThreshold    *float64       `json:"mcid_threshold,omitempty"`
	MCIDAchieved     bool           `json:"mcid_achieved"`
	Interpretation   Interpretation `json:"interpretation"`
}

// ReevaluationSummary groups re-evaluation results for a patient visit.
type ReevaluationSummary struct {
	PatientID    string                    `json:"patient_id"`
	VisitID      *string                   `json:"visit_id,omitempty"`
	TherapistID  string                    `json:"therapist_id"`
	AssessedAt   time.Time                 `json:"assessed_at"`
	Comparisons  []ReevaluationAssessment  `json:"comparisons"`
	TotalItems   int                       `json:"total_items"`
	Improved     int                       `json:"improved"`
	Declined     int                       `json:"declined"`
	Stable       int                       `json:"stable"`
	MCIDAchieved int                       `json:"mcid_achieved"`
}

// CreateReevaluationRequest represents the request body for creating re-evaluation records.
type CreateReevaluationRequest struct {
	PatientID            string                         `json:"patient_id" validate:"required,uuid"`
	VisitID              string                         `json:"visit_id" validate:"omitempty,uuid"`
	BaselineAssessmentID string                         `json:"baseline_assessment_id" validate:"omitempty,uuid"`
	Assessments          []CreateReevaluationItemRequest `json:"assessments" validate:"required,min=1"`
	Notes                string                         `json:"notes" validate:"max=2000"`
	AssessedAt           string                         `json:"assessed_at" validate:"omitempty,datetime=2006-01-02T15:04:05Z07:00"`
}

// CreateReevaluationItemRequest represents a single comparison item in a re-evaluation.
type CreateReevaluationItemRequest struct {
	AssessmentType AssessmentType `json:"assessment_type" validate:"required"`
	MeasureLabel   string         `json:"measure_label" validate:"required,max=120"`
	CurrentValue   float64        `json:"current_value"`
	BaselineValue  float64        `json:"baseline_value"`
	HigherIsBetter bool           `json:"higher_is_better"`
	MCIDThreshold  *float64       `json:"mcid_threshold,omitempty"`
}

// CalculateChange computes the change and percentage between baseline and current.
func CalculateChange(baseline, current float64) (change float64, changePct *float64) {
	change = current - baseline
	if math.Abs(baseline) > 0.0001 {
		pct := (change / math.Abs(baseline)) * 100
		changePct = &pct
	}
	return change, changePct
}

// DetermineInterpretation determines if the change represents improvement, decline, or stability.
// mcid is the Minimal Clinically Important Difference threshold.
// higherIsBetter indicates whether higher values represent improvement (true for ROM/function,
// false for pain scales).
func DetermineInterpretation(change float64, mcid *float64, higherIsBetter bool) (Interpretation, bool) {
	threshold := 0.0
	if mcid != nil && *mcid > 0 {
		threshold = *mcid
	}

	absChange := math.Abs(change)

	// If there is a meaningful MCID and the change does not exceed it, consider stable
	if threshold > 0 && absChange < threshold {
		return InterpretationStable, false
	}

	// If there is no MCID but the change is essentially zero, consider stable
	if threshold == 0 && absChange < 0.0001 {
		return InterpretationStable, false
	}

	mcidAchieved := threshold > 0 && absChange >= threshold

	if higherIsBetter {
		if change > 0 {
			return InterpretationImproved, mcidAchieved
		}
		return InterpretationDeclined, mcidAchieved
	}

	// For pain scales where lower is better
	if change < 0 {
		return InterpretationImproved, mcidAchieved
	}
	return InterpretationDeclined, mcidAchieved
}
